File: library/src/main/java/com/snail/commons/base/BaseListAdapter.java
package com.snail.commons.base;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseListAdapter<T> extends BaseAdapter {
	private Context context;
	private List<T> data;

	public BaseListAdapter(@NonNull Context context) {
		data = new ArrayList<>();
		this.context = context;
	}
	
	public BaseListAdapter(@NonNull Context context, @NonNull List<T> data) {
		this.context = context;
		this.data = data;
	}

	public List<T> getData() {
		return data;
	}

	public void setData(@NonNull List<T> data) {
		this.data = data;
		notifyDataSetChanged();
	}
	
	public Context getContext() {
        return context;
    }
	
	@Override
	public int getCount() {
		return data.size();
	}

	@Override
	public T getItem(int position) {
		return data.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		BaseHolder<T> holder;
		if (convertView == null) {
			holder = getHolder(position);
		} else {
			holder = (BaseHolder<T>) convertView.getTag();
		}
		holder.setData(data.get(position), position);
		return holder.getConvertView();
	}

	protected abstract BaseHolder<T> getHolder(int position);
}

File: library/src/main/java/com/snail/commons/base/BaseListAdapter.kt
package com.snail.commons.base

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import java.util.*
import kotlin.collections.HashMap

abstract class BaseListAdapter<T> @JvmOverloads constructor(val context: Context, list: MutableList<T> = ArrayList()) : BaseAdapter() {
    private val holders = HashMap<View, BaseViewHolder<T>>()
    
    var items: MutableList<T> = list
        set(value) {
            field = value
            notifyDataSetChanged()
        }
        
    fun refresh(list: List<T>) {
        items.clear()
        items.addAll(list)
        notifyDataSetChanged()
    }

    override fun notifyDataSetChanged() {
        holders.clear()
        super.notifyDataSetChanged()
    }
    
    override fun getCount(): Int {
        return items.size
    }

    override fun getItem(position: Int): T {
        return items[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val viewHolder: BaseViewHolder<T> = if (convertView == null) {
            val h = createViewHolder(position)
            holders[h.convertView] = h
            h
        } else {
            holders[convertView]!!
        }
        viewHolder.onBind(items[position], position)
        return viewHolder.convertView
    }

    protected abstract fun createViewHolder(position: Int): BaseViewHolder<T>
}